deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/preprocessor/Cnn3DToFeedForwardPreProcessor.java

```diff
@@ -0,0 +1,163 @@
+ * Note: numChannels is equivalent to featureMaps referenced in different literature
+ *
+ * @author Max Pumperla
+ * @see FeedForwardToCnn3DPreProcessor for opposite case (i.e., DenseLayer -> CNN3D)
+ */
+@Data
+public class Cnn3DToFeedForwardPreProcessor implements InputPreProcessor {
+    protected int inputDepth;
+    protected int inputHeight;
+    protected int inputWidth;
+    protected int numChannels;
+    protected boolean isNCDHW = true; // channels first ordering by default
+
+    /**
+     * @param inputDepth  input channels
+     * @param inputHeight input height
+     * @param inputWidth  input width
+     * @param numChannels input channels
+     * @param isNCDHW     boolean to indicate data format, i.e. channels first (NCDHW) vs. channels last (NDHWC)
+     */
+    @JsonCreator
+    public Cnn3DToFeedForwardPreProcessor(@JsonProperty("inputDepth") int inputDepth,
+                                          @JsonProperty("inputHeight") int inputHeight,
+                                          @JsonProperty("inputWidth") int inputWidth,
+                                          @JsonProperty("numChannels") int numChannels,
+                                          @JsonProperty("isNCDHW") boolean isNCDHW) {
+        this.inputDepth = inputDepth;
+        this.inputHeight = inputHeight;
+        this.inputWidth = inputWidth;
+        this.numChannels = numChannels;
+        this.isNCDHW = isNCDHW;
+    }
+
+    public Cnn3DToFeedForwardPreProcessor(int inputDepth, int inputHeight, int inputWidth) {
+        this.inputDepth = inputDepth;
+        this.inputHeight = inputHeight;
+        this.inputWidth = inputWidth;
+        this.numChannels = 1;
+    }
+
+    public Cnn3DToFeedForwardPreProcessor() {
+    }
+
+    @Override
+    public INDArray preProcess(INDArray input, int miniBatchSize) {
+        if (input.rank() == 2)
+            return input; // Pass-through feed-forward input
+
+        // We expect either NCDHW or NDHWC format
+        if ((isNCDHW && input.size(1) != numChannels) || (!isNCDHW && input.size(4) != numChannels)) {
+            throw new IllegalStateException("Invalid input array: expected shape in format "
+                    + "[minibatch, channels, channels, height, width] or "
+                    + "[minibatch, channels, height, width, channels]"
+                    + "for numChannels: " + numChannels + ", inputDepth " + inputDepth + ", inputHeight " + inputHeight
+                    + " and inputWidth " + inputWidth + ", but got "
+                    + Arrays.toString(input.shape()));
+        }
+
+
+        //Assume input is standard rank 5 activations out of CNN3D layer
+        //First: we require input to be in c order. But c order (as declared in array order) isn't enough;
+        // also need strides to be correct
+        if (input.ordering() != 'c' || !Shape.strideDescendingCAscendingF(input))
```

Comment:
These days: I'm using this method instead of strideDescending... as it's a bit more strict (and catches more invalid cases)
https://github.com/deeplearning4j/nd4j/blob/5d4e8a0a873eed92980910552767ed2626d1270b/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/shape/Shape.java#L2609-L2622